[zackjh] iP

zackjh opened the pull request:
# Evan

> "The biggest constraint of the next 100 years of computing is the idea of metaphors." - Evan Spiegel, CEO of Snap Inc. ([source](https://sg.finance.yahoo.com/news/11-brilliant-quotes-evan-spiegel-191024037.html?guccounter=1))

Evan frees your mind of having to remember things you need to do. It's:

- text-based
- intuitive
- ~~fast~~ really fast

All you need to do is:

1. Download it from [here](https://github.com/zackjh/ip/releases)
2. Double-click it
3. Add your tasks
4. Reap the benefits of using Evan

Best of all, it's **free**! 😁

Features:

- [x] Managing tasks (todos, deadlines, and events)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practise your Java skills too! Here's the `main` method:
```java
public static void main(String[] args) {
    new Evan("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zackjh committed:
Rename chatbot to Evan

zackjh committed:
Update Evan so that it echoes the user's input

zackjh committed:
Allow user to store tasks

zackjh committed:
Refactor Evan.java by extracting the code in main() into their own functions

zackjh committed:
Create Task class and use it to represent a task

zackjh committed:
Add ability to mark tasks as done

zackjh committed:
Add support for tracking three types of tasks

zackjh committed:
Semi-automate the testing of Evan using input/output redirection

zackjh committed:
Allow Evan to handle invalid user inputs

zackjh committed:
Add support for deleting tasks from the list

zackjh committed:
Handle invalid task numbers

zackjh committed:
Refactor code to use more OOP and prepare for storage feature

zackjh committed:
Implement loading of tasks from tasks.txt

zackjh committed:
Implement encodeAsString method for Tasks

zackjh committed:
Save tasks to data/tasks.txt automatically

zackjh committed:
Merge branch 'branch-Level-7'

zackjh committed:
Store task datetimes as custom DateTime objects

zackjh committed:
Merge branch 'branch-Level-8'
evan.task.DateTime was created so that the start/end date/time for Task objects can be represented more accurately.

zackjh committed:
Merge branch 'branch-A-Gradle'
Merged as part of A-Gradle task.

zackjh committed:
Update mainClass to point to correct class

zackjh committed:
Add unit tests for DateTime class

zackjh committed:
Make test comments more descriptive

zackjh committed:
Add unit tests for Todo, Deadline, and Event

zackjh committed:
Update shadowJar config in build.gradle

zackjh committed:
Add JavaDoc header comments to all non-test classes

zackjh committed:
Unindent 'case' branches in 'switch' statements

zackjh committed:
Remove unnecessary blank line

zackjh committed:
Add 'find' command to UI's list of commands

zackjh committed:
Override toString() in TaskList

zackjh committed:
Implement 'find' command

zackjh committed:
Merge branch 'branch-A-JavaDoc'

zackjh committed:
Merge branch 'branch-A-CodingStandard'

zackjh committed:
Merge branch 'branch-Level-9'

zackjh committed:
Add JavaDoc comments to classes in the 'command' package

zackjh committed:
Add JavaDoc comment to 'toString' method in TaskList

zackjh committed:
Add JavaDoc comments to files in the 'exception' package

zackjh committed:
Reformat code and optimize imports

zackjh committed:
Add checkstyle config files

zackjh committed:
Update checkstyle config files to match those in AddressBook Level 3

zackjh committed:
Update build.gradle to support checkstyle

zackjh committed:
Fix checkstyle errors for main files

zackjh committed:
Fix checkstyle errors for test files

zackjh committed:
Merge branch 'branch-A-CheckStyle'
Set up checkstyle and updated source files that violated checkstyle rules.

zackjh committed:
Rename 'main' package to 'service' and configure gradle for JavaFX

raageth commented on src/main/java/evan/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,43 @@
+        this.by = by;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
```
I like how using `{@inheritDoc}` keeps the code clean.

raageth commented on src/main/java/evan/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+     * Returns true if the chatbot should exit the session after execution of the command, and false if the chatbot
+     * should continue to prompt the user for input.
+     */
+    public abstract boolean isExit();
```
Since `ExitCommand` is the only place where `isExit()` returns `true`,  why not set `isExit()` to return `false` here as a default value?

colinhia replied:
Adding on to the point above, perhaps another approach is instead of a method, you could consider using a boolean field instantiated to be false by default? Then you could just access the field instead of having to implement the method everywhere

raageth commented on src/main/java/evan/main/Storage.java:
```diff
@@ -0,0 +1,134 @@
+     * @return ArrayList of Tasks containing tasks decoded from the storage .txt file.
+     * @throws LoadingException If there is an error reading the storage .txt file.
+     */
+    public ArrayList<Task> load() throws LoadingException {
```
Could a more specific method name be used in this case?

raageth commented on src/main/java/evan/task/DateTime.java:
```diff
@@ -0,0 +1,183 @@
+     * @return True if the DateTime object is using a STRING, and false otherwise.
+     */
+    public boolean isString() {
+        return type == InputType.STRING;
```
Would using `type.equals(InputType.STRING)` be better?

raageth commented on src/test/java/evan/task/DateTimeTest.java:
```diff
@@ -0,0 +1,32 @@
+public class DateTimeTest {
+    @Test
+    public void parseInput_variousString_success() {
+        // Simple string → STRING
+        assertEquals(DateTime.InputType.STRING, DateTime.parseInput("buy bread").type);
+
+        // Invalid date → STRING
+        assertEquals(DateTime.InputType.STRING, DateTime.parseInput("2024-13-02").type);
+
+        // Valid date, invalid time → STRING
+        assertEquals(DateTime.InputType.STRING, DateTime.parseInput("2024-11-02 2501").type);
+
+        // Invalid date, invalid time → STRING
+        assertEquals(DateTime.InputType.STRING, DateTime.parseInput("2024-12-32 1261").type);
+
+        // Invalid date, valid time → STRING
+        assertEquals(DateTime.InputType.STRING, DateTime.parseInput("2024-00-25 1200").type);
+
+        // Valid date → DATE
+        assertEquals(DateTime.InputType.DATE, DateTime.parseInput("2024-04-06").type);
+
+        // Valid date and time → DATETIME
+        assertEquals(DateTime.InputType.DATETIME, DateTime.parseInput("2024-09-10 2359").type);
```
I like how thorough the testing is.

raageth reviewed (commented):
LGTM. The code was well written and I could not find much coding standard issues.

colinhia commented on src/main/java/evan/command/TodoCommand.java:
```diff
@@ -0,0 +1,40 @@
+ * Represents a command that the chatbot can execute to create a Todo.
+ */
+public class TodoCommand extends Command {
+    private final String description;
```
I noticed that the String description is common to the Todo/Deadline/Event commands, perhaps there could be another abstract class grouping these commands for greater encapsulation?

colinhia commented on src/main/java/evan/main/Evan.java:
```diff
@@ -0,0 +1,72 @@
+        }
+        System.exit(0);
+    }
+}
```
I liked how specific your exceptions were! Perhaps one suggestion I would make is have them grouped under an abstract EvanException class? Then you might be able to handle the exceptions more easily in the main Evan class.

colinhia commented on src/main/java/evan/main/Storage.java:
```diff
@@ -0,0 +1,134 @@
+            case "E" -> {
+                String from = parts[3];
+                String to = parts[4];
+                task = new Event(description, from, to);
```
I liked the use of the switch statement here!

colinhia commented on src/main/java/evan/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+    public String toString() {
+        return String.format("[%s] %s", this.getStatusIcon(), this.description);
+    }
+}
```
I liked the grouping of common methods into the overarching abstract task class!

colinhia reviewed (commented):
Overall great job! I liked how consistent the coding style was throughout the project, from the import statements, to the indentation in the code. I think it is no easy feat to maintain that, especially with the use of switch statements and the such. Keep it up!

zackjh committed:
Set up basic JavaFX application

zackjh committed:
Implement GUI

zackjh committed:
Merge branch 'branch-Level-10'
Set up GUI for Evan chatbot.

zackjh committed:
Add assertions to check for non-null instance variables

zackjh committed:
Add assertion to check for non-null result in toString method

zackjh committed:
Add comments to Storage.java

zackjh committed:
Add comments to Evan.java

zackjh committed:
Merge branch-A-Assertions to master

Add assertions to codebase to improve reliability

zackjh committed:
Merge branch-A-CodeQuality into master

Improve code quality of codebase by adding comments

zackjh committed:
Add 'help' command to access a help page

zackjh committed:
Merge branch 'branch-C-Help'

'branch-C-Help' implemented the 'help' command which provides users
with guidance on how to use the chat bot.

zackjh committed:
Update README to remove mentions of 'Duke'

There were many mentions of 'Duke' in the README which have since been
removed.

The README is incomplete - more work has to be done to craft a
functional and informative README.

zackjh committed:
Change images for User and Evan

zackjh committed:
Remove unnecessary comment

zackjh committed:
Add Ui screenshot to docs

zackjh committed:
Update user guide

zackjh committed:
Update README

zackjh committed:
Use dynamic filepaths instead of hardcoded filepaths
